core/src/test/scala/unit/kafka/server/KafkaApisTest.scala

```diff
@@ -9871,6 +9871,69 @@ class KafkaApisTest extends Logging {
     assertEquals(Errors.GROUP_AUTHORIZATION_FAILED.code, response.data.errorCode)
   }
 
+  @Test
+  def testConsumerGroupHeartbeatFilterUnauthorizedTopics(): Unit = {
+    val memberId = Uuid.randomUuid.toString
+    val fooTopicName = "foo"
+    val barTopicName = "bar"
+    val fooTopicId = Uuid.randomUuid
+    val barTopicId = Uuid.randomUuid
+    val zarTopicId = Uuid.randomUuid
+
+    metadataCache = mock(classOf[KRaftMetadataCache])
+    when(metadataCache.getTopicName(fooTopicId)).thenReturn(Some(fooTopicName))
+    when(metadataCache.getTopicName(barTopicId)).thenReturn(Some(barTopicName))
+    when(metadataCache.getTopicName(zarTopicId)).thenReturn(None)
+
+    val consumerGroupHeartbeatRequest = new ConsumerGroupHeartbeatRequestData().setGroupId("group")
+
+    val requestChannelRequest = buildRequest(new ConsumerGroupHeartbeatRequest.Builder(consumerGroupHeartbeatRequest).build())
+
+    val future = new CompletableFuture[ConsumerGroupHeartbeatResponseData]()
+    when(groupCoordinator.consumerGroupHeartbeat(
+      requestChannelRequest.context,
+      consumerGroupHeartbeatRequest
+    )).thenReturn(future)
+
+    val authorizer: Authorizer = mock(classOf[Authorizer])
+    def buildExpectedActions(topics: List[String]): util.List[Action] = {
+      topics.map { topic =>
+        val pattern = new ResourcePattern(ResourceType.TOPIC, topic, PatternType.LITERAL)
+        new Action(AclOperation.DESCRIBE, pattern, 1, true, true)
+      }.asJava
+    }
+    when(authorizer.authorize(any[RequestContext], any[util.List[Action]]))
+      .thenReturn(Seq(AuthorizationResult.ALLOWED).asJava)
+    when(authorizer.authorize(any[RequestContext], ArgumentMatchers.eq(buildExpectedActions(List(fooTopicName, barTopicName)))))
+      .thenReturn(List(AuthorizationResult.ALLOWED, AuthorizationResult.DENIED).asJava)
+    when(authorizer.authorize(any[RequestContext], ArgumentMatchers.eq(buildExpectedActions(List(barTopicName, fooTopicName)))))
+      .thenReturn(List(AuthorizationResult.DENIED, AuthorizationResult.ALLOWED).asJava)
```

Comment:
I recall using a nice pattern to do this. Take a look into `testHandleOffsetFetchAuthorization`. It would avoid having to specify all the combinations.